src/h2load.cc

```diff
@@ -252,6 +260,56 @@ void rate_period_timeout_w_cb(struct ev_loop *loop, ev_timer *w, int revents) {
 } // namespace
 
 namespace {
+  // Called when the warmup duration for infinite number of requests are over
+  void warmup_timeout_cb(struct ev_loop *loop, ev_timer *w, int revents) {
+    auto client = static_cast<Client *>(w->data);
+    
+    if (client->worker->current_phase != MAIN_DURATION) {
+      client->worker->current_phase = MAIN_DURATION;
+      std::cout << "Warm-up phase is over for client: " << client->id << std::endl;
+      std::cout << "Main benchmark duration is started." << std::endl;
+    }
+    
+    assert (client->worker->stats.req_started == 0);
+    assert (client->worker->stats.req_done == 0);
+    // client->worker->stats.req_started -= client->req_started;
+    // client->worker->stats.req_done -= client->req_done;
+
+    assert (client->req_todo == 0);
+    assert (client->req_left == 1);
+    assert (client->req_inflight == 0);
+    assert (client->req_started == 0);
+    assert (client->req_done == 0);
+    
+    client->record_client_start_time();
+    client->clear_connect_times();
+    client->record_connect_start_time();
+    
+    ev_timer_start(client->worker->loop, &client->duration_watcher);
+    
+    return;
+  }
+}
+
+namespace {
+  // Called when the duration for infinite number of requests are over
+  void duration_timeout_cb(struct ev_loop *loop, ev_timer *w, int revents) {
+    auto client = static_cast<Client *>(w->data);
+
+    client->req_todo = client->req_done; // there was no finite "req_todo"
+    client->worker->stats.req_todo += client->req_todo;
+    client->req_inflight = 0;
+    client->req_left = 0;
+    client->measurement_calculation_done = true;
+    
+    if (client->worker->current_phase != DURATION_OVER) {
```

Comment:
Yeah, the same logic. The first client sets that the `DURATION_OVER`, then rest of the clients do not enter into the `if` block.